rs/https_outcalls/consensus/src/gossip.rs

```diff
@@ -10,9 +10,17 @@ use ic_interfaces::{
 use ic_interfaces_state_manager::StateReader;
 use ic_logger::{warn, ReplicaLogger};
 use ic_replicated_state::ReplicatedState;
-use ic_types::{artifact::CanisterHttpResponseId, canister_http::CanisterHttpResponseShare};
+use ic_types::{
+    artifact::CanisterHttpResponseId, canister_http::CanisterHttpResponseShare,
+    messages::CallbackId,
+};
 use std::{collections::BTreeSet, sync::Arc};
 
+/// The upper bound of how many shares of HTTP outcall requests unknown to the local replica will be
+/// requested via P2P in advance. Since the request ids are strictly increasing, we can simply add
+/// the constant to the latest known request id.
+const MAX_NUMBER_OF_REQUESTS_AHEAD: u64 = 10;
```

Comment:
Where does this number come from? I guess it should be related to the maximum increase in callback IDs we expect to see within the 3 seconds update frequency of the priority function

Reply:
I just started with something, expecting that we can discuss an appropriate value during the review. The problem is that since it ultimately depends on the number of canisters on the subnet, no constant is going to work in all cases. `10` sounds like it should work with the current load, but we can go with 100 as well. 

Any preference?

Reply:
I guess the suggestion in the [https://dfinity.atlassian.net/browse/CON-1229](ticket) was to first come up with some reasonable bound for the maximum number of requests that the subnet will work on concurrently (this is also needed to bound the pool size in general). With that we could then come up with some bound on the maximum load we might see.

However the problem is that these callback IDs are not only incremented by http outcalls, so this would still just be an heuristic. The good news is that `MAX_NUMBER_OF_REQUESTS_AHEAD` is really only needed for performance reasons, in order to work around the fact that `known_request_ids` is only updated every three seconds. So even if we set the constant to 0, it should still work.

Are you aware of any metrics for `subnet_call_context_manger.next_callback_id()`? That could help us figure out by how much these IDs currently increase within 3 seconds.

Reply:
I think you linked the wrong ticket, in the [correct](https://dfinity.atlassian.net/browse/CON-1227) ticket there are no suggestions.

I checked the HTTP rates on a couple of subnets including the largest ones and the http request rate from the last 30 days was never above 1 r/sec.

But `next_callback_id` can be increased by other things as well. I pinged the execution team.

Reply:
@eichhorl so the [best approximation](https://grafana.mainnet.dfinity.network/explore?orgId=1&left=%7B%22datasource%22:%22P119963096DF10E67%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P119963096DF10E67%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%20%28ic_subnet%29%20%28%5Cn%20%20quantile%20by%28method_name,%20ic_subnet%29%20%28%5Cn%20%20%20%200.5,%5Cn%20%20%20%20sum%20without%28outcome,%20status%29%20%28%5Cn%20%20%20%20%20%20rate%28execution_subnet_message_duration_seconds_count%7Bjob%3D%5C%22replica%5C%22,%20ic%3D%5C%22mercury%5C%22,%20method_name%3D~%5C%22ic00_http_request%5C%22%7D%5B10s%5D%29%5Cn%20%20%20%20%29%5Cn%20%20%29%5Cn%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%22now-30d%22,%22to%22:%22now%22%7D%7D) I could extract with the help of the execution team is the duration of http outcalls on all subnets within a 10s interval in the last 30 days. It always always stays around 2.5. So the constant `10` seems to me like it's still more than enough. We could also reduce it, but I am not sure if it would have any effects. WDYT?

Reply:
What do you mean by "duration"? If I press "show all 23" and choose 0.95 as the quantile then there are some spikes above 15. Also, should we count all methods that can increase the callback ID? I think the only downside of choosing a slightly larger constant would be a slightly bigger unvalidated pool for a while. So maybe ~50 could give us enough runway for now?

Reply:
Here's the [query](https://grafana.mainnet.dfinity.network/explore?orgId=1&left=%7B%22datasource%22:%22P119963096DF10E67%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P119963096DF10E67%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%20%28ic_subnet%29%20%28%5Cn%20%20quantile%20by%28method_name,%20ic_subnet%29%20%28%5Cn%20%20%20%200.95,%5Cn%20%20%20%20sum%20without%28outcome,%20status%29%20%28%5Cn%20%20%20%20%20%20rate%28execution_subnet_message_duration_seconds_count%7Bjob%3D%5C%22replica%5C%22,%20ic%3D%5C%22mercury%5C%22,%20method_name%3D~%5C%22ic00_bitcoin_get_successors%7Cic00_bitcoin_send_transaction_internal%7Cic00_compute_initial_i_dkg_dealings%7Cic00_http_request%7Cic00_setup_initial_dkg%7Cic00_sign_with_ecdsa%7Cic00_sign_with_schnorr%5C%22%7D%5B10s%5D%29%5Cn%20%20%20%20%29%5Cn%20%20%29%5Cn%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%22now-30d%22,%22to%22:%22now%22%7D%7D) for 0.95 quantile and all methods. There are only 2 spikes in 30 days: around 25 and 30 requests. So IMO `50` would work 👍

Reply:
Sounds good!